botocore/client.py

```diff
@@ -175,14 +180,38 @@ def _get_s3_addressing_handler(self, endpoint_url, s3_config,
         return fix_s3_host
 
     def _is_s3_accelerate(self, endpoint_url, s3_config):
+        # Accelerate has been explicitly configured
         if s3_config is not None and s3_config.get('use_accelerate_endpoint'):
             return True
 
+        # Accelerate mode is turned on automatically if an endpoint url is
+        # provided that matches the accelerate scheme.
         if endpoint_url is None:
             return False
 
-        parts = urlsplit(endpoint_url).netloc.split('.')
-        return any(part == 's3-accelerate' for part in parts)
+        # Accelerate is only valid for Amazon endpoints.
+        netloc = urlsplit(endpoint_url).netloc
+        if not netloc.endswith('amazonaws.com'):
+            return False
+
+        # The first part of the url should always be s3-accelerate
```

Comment:
I do not think this needs to change, but I had to double check closely that there would be no indexing errors. I think you could simplify with a length check base approach:

```python
parts = netloc.split('.')

if len(parts) not in [3, 4]:
    return False

if parts[0] != 's3-accelerate':
    return False

if len(parts) == 4:
    return parts[1] in ['dualstack']

return True
```

Feel free to use it, but I think what you are doing has no missing edge cases so it should be fine.